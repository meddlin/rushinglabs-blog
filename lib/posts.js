import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';
import { serialize } from 'next-mdx-remote/serialize'


const postsDirectory = path.join(process.cwd(), 'posts');
const docsDirectory = path.join(process.cwd(), 'docs');


/**
 * Recursively walks the /posts directory to find all posts. Returns an array files; full file path, not just file name.
 * @param {*} dirPath 
 * @param {*} arrayOfFiles 
 */
const getAllFiles = function(dirPath, arrayOfFiles) {
    const files = fs.readdirSync(dirPath);
    arrayOfFiles = arrayOfFiles || [];

    // NOTE: The '/' character technically works on Windows, but likely because of how loose Windows can be with
    // file paths, for easier use. HOWEVER, switching to '\\' to use a backslash...for a true system path the way it
    // should be...ALSO works on Windows. The '\' char needs an escape, so '\\' is what we're left with.
    files.forEach(function(file) {
        if (fs.statSync(dirPath + "/" + file).isDirectory())
            arrayOfFiles = getAllFiles(dirPath + "/" + file, arrayOfFiles);
        else
            arrayOfFiles.push(path.join(dirPath, "/", file));
    });

    return arrayOfFiles;
};


/**
 * Returns a list of file path objects (all blog posts)
 * -- { file: 'path', year: 'year-subdirectory' }
 * 
 * Recursively walks the /posts directory to find all posts--like `getAllFiles()`, but returns full file & year info.
 * @param {*} dirPath 
 * @param {*} arrayOfFiles 
 */
const getAllFilesWithYearInfo = function(dirPath) {
    const filePaths = getAllFiles(dirPath);
    let fileInfo = [];

    // Full file paths were returned in `filePaths`
    // Iterate over each one to pull the last 2 pieces of info off of the path
    filePaths.map(file => {
        // ***
        // NOTE: Does splitting on the file path delimiter work across multiple operating systems? (i.e. file systems)
        // ***
        let splits = file.split(path.sep);

        fileInfo.push({
            file: splits[splits.length - 1],
            year: splits[splits.length - 2],
        })
    });

    return fileInfo;
}


/**
 * Produce a unique array of all categories mentioned across all posts.
 * 
 * Locations used
 * - pages/index.js: used to list all categories on the landing page
 */
export function getAllCategories() {
    const postInfo = getAllFilesWithYearInfo(postsDirectory);

    const allCategories = postInfo.map(post => {
        // Rejoin the `postsDirectory`, `post year`, and `post filename` to build the file path
        const fullPath = path.join(postsDirectory, `/${post.year}/${post.file}`);
        // Use file path to read the file
        const fileContents = fs.readFileSync(fullPath, 'utf8');
        // Use `matter()` to pull the "matter info" from the top of the markdown file
        const matterResult = matter(fileContents);

        // Pull the `section` descriptor out of the markdown info at the top of the blog post
        return (matterResult && matterResult.data && matterResult.data.section) ? matterResult.data.section : '';
    });

    const uniqueCategories = [...new Set(allCategories)];
    return uniqueCategories;
};

/**
 * 
 * @param {*} directory 
 * @returns 
 */
export function getPostsFromDocsSubdir(directory) {
    const postsArr = getAllFiles(`${docsDirectory}/${directory}`);
    let docsSubdirPosts = [];

    for(let i = 0; i < postsArr.length; i++) {
        let splits = postsArr[i].split(path.sep);

        const id = splits[splits.length - 1].replace(/\.mdx$/, '');

        const fileContents = fs.readFileSync(postsArr[i], 'utf8');
        const matterResult = matter(fileContents);

        docsSubdirPosts.push({ id, directory, ...matterResult.data });
    }

    return docsSubdirPosts;
}


/**
 * Return a list of posts containing a specific category tag
 * @param {*} category 
 */
export function getCategoryPosts(category) {
    const postsInfo = getAllFilesWithYearInfo(postsDirectory);
    let allCategoryPosts = [];

    for(let i = 0; i < postsInfo.length; i++) {
        const id = postsInfo[i].file.replace(/\.mdx$/, '');
        const year = postsInfo[i].year;
        const fullPath = path.join(postsDirectory, `/${postsInfo[i].year}/${postsInfo[i].file}`);
        const fileContents = fs.readFileSync(fullPath, 'utf8');
        const matterResult = matter(fileContents);

        if ((matterResult && matterResult.data) && matterResult.data.section === category) {
            allCategoryPosts.push({ id, year, ...matterResult.data });
        }
    }

    // Sort posts by date
    return allCategoryPosts.sort((a, b) => {
        if (a.date < b.date) return 1;
        else return -1;
    });
}



///
// Fetch a specific post's data
///
export async function getPostData(year, id) {
    // TODO: Support .md files
    const fullPath = path.join(postsDirectory, `${year}/${id}.mdx`);
    const markdownWithMeta = fs.readFileSync(fullPath, 'utf8');

    // use gray-matter to parse the post metadata section
    const frontMatter = matter(markdownWithMeta);

    // frontMatter.content is the MD post
    const mdxSource = await serialize(frontMatter.content);

    // Combine the data with the id
    return {
        id,
        mdxSource,
        ...frontMatter.data
    };
}



///
// Retrieve a list of file names in the `/posts` directory (excluding '.md' extension from name) 
///
export function getAllPostIds() {
    let fileNames = getAllFiles(postsDirectory);
    var postInfo = [];

    for(var i = 0; i < fileNames.length; i++) {
        let noExt;
        /**
         * Support for Markdown (.md) and MDX (.mdx) files
         */
        
        // TODO : Support .md files
        if (fileNames[i].endsWith('.mdx')) noExt = fileNames[i].replace(/\.mdx$/, '');

        let splitPath = noExt.split(path.sep);
        let fileName = splitPath[splitPath.length - 1];
        let year = splitPath[splitPath.length - 2];

        postInfo.push({
            params: {
                id: fileName,
                year: year
            }
        });
    }

    return postInfo;    
};


/**
 * Returns a sorted list of all posts
 * This is the main logic used 
 * 
 * Locations used
 * - pages/index.js: Grabs all of the posts
 */
export function getSortedPostsData() {
    const fileNames = getAllFiles(postsDirectory); // Filenames contain the full path

    /**
     * TODO : Move the filename replacement, year extraction, and frontmatter metadata parsing
     * to another function. This would allow the .filter().map() functionality, to direct the 
     * decision for what gets passed along, be MUCH easier to read.
     * 
     * Also, likely makes this logic a bit more testable. This is some core functionality to 
     * the site.
     */
    const allPostsData = fileNames.filter(fileName => {
        // Remove the '.md' file extension, and grab the last section of the file path to...
        // Create an id from the filename
        var id = fileName.replace(/\.mdx$/, '');
        id = id.split(path.sep)[id.split(path.sep).length - 1];

        const year = fileName.split(path.sep)[fileName.split(path.sep).length - 2];
        id = `${year}/${id}`;

        // Read whole markdown file as string
        const fileContents = fs.readFileSync(fileName, 'utf8');

        // Use gray-matter to parse the post metadata section
        const matterResult = matter(fileContents);

        return matterResult.data.published == true ? true : false; // Wanting to be explicit here.
    }).map(fileName => {
        var id = fileName.replace(/\.mdx$/, '');
        id = id.split(path.sep)[id.split(path.sep).length - 1];
        const year = fileName.split(path.sep)[fileName.split(path.sep).length - 2];
        id = `${year}/${id}`;
        const fileContents = fs.readFileSync(fileName, 'utf8');
        const matterResult = matter(fileContents);
        // console.log('matterResult ---');
        // console.log(matterResult);

        return {
            id,
            ...matterResult.data
        }
    });

    // Sort posts by date
    return allPostsData.sort((a, b) => {
        if (a.date < b.date) {
            return 1;
        } else {
            return -1;
        }
    });
};
