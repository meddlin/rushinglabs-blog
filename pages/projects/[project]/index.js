import escapeHTML from 'escape-html';
import Layout from '../../../components/layout';
import { getPostsFromProjectsSubdir, getAllProjects } from '../../../lib/projects';

export async function getStaticProps({ params }) {
    const projectPosts = getPostsFromProjectsSubdir(params.project);

    return {
        props: {
            projectTitle: params.project,
            posts: projectPosts
        }
    }
}

export async function getStaticPaths() {
    const projects = getAllProjects();

    return {
        paths: projects.map(proj => {
            return {
                params: { project: proj }
            }
        }),
        fallback: false
    }
}

const ProjectsSection = ({ projectTitle, posts }) => {
    return (
        <Layout>
            <h1>Project: {projectTitle}</h1>

            {posts ? (
                <ul>
                {
                    posts.map( post => (
                        <li key={post.title}>
                            <a href={`/projects/${escapeHTML(projectTitle)}/${escapeHTML(post.id)}`}>
                                {post.title}
                            </a>
                        </li>
                    ))
                }
                </ul>
            ) : ''}
        </Layout>
    );
}

export default ProjectsSection;