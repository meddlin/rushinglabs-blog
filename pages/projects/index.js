import Layout, { siteTitle } from '../../components/layout';
import Head from 'next/head';
import Image from 'next/image';
import SocialsHorizontal from '../../components/socials-horizontal';
import utilStyles from '../../styles/utils.module.css';
import styles from '../../styles/projects.module.css';
import { getAllProjects } from '../../lib/projects';
import { projectsMeta } from '../../projectsConfig';

/**
 *
 * /projects/index.js
 * The "view all" view for projects. Should be able to find
 * all projects from this view.
 *
 */

export async function getStaticProps() {
	const projects = getAllProjects();

	const projectMetadata = projectsMeta;

	return {
		props: {
			projects: projects,
			metadata: projectMetadata,
		},
	};
}

const Projects = ({ metadata, projects }) => {
	return (
		<Layout footerDisplay={true}>
			<Head>
				<title>{siteTitle}</title>
			</Head>

			<section
				className={`${utilStyles.headingMd} ${utilStyles.padding1px} ${utilStyles.postsSection}`}
			>
				<ul className={styles.listParent}>
					{metadata && metadata.length > 0
						? metadata.map(meta =>
								meta.published ? (
									<li key={meta.title}>
										<div style={{ display: `flex`, flexDirection: `row` }}>
											<div style={{ maxWidth: `65%` }}>
												<p>
													<a href={`/projects/${meta.title}`}>{meta.title}</a>
												</p>
												<p>{meta.summary}</p>
												<p className={styles.projectLinks}>
													{meta.links.map(link => (
														<a href={link.url}>{link.text}</a>
													))}
												</p>
											</div>
											{meta.images &&
											(meta.images.length && meta.images[0].url !== '') >
												0 ? (
												<Image
													src={meta.images[0].url}
													alt={meta.images[0].alt}
													height={meta.images[0].height}
													width={meta.images[0].width}
												/>
											) : (
												''
											)}
										</div>
									</li>
								) : (
									''
								),
						  )
						: ''}
				</ul>
			</section>
		</Layout>
	);
};

export default Projects;
