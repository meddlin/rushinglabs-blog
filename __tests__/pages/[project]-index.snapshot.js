import { render } from '@testing-library/react';
import ProjectsSection from '../../pages/projects/[project]/index';
import { getPostsFromProjectsSubdir, getAllProjects } from '../../lib/projects';

/**
 * Since projects are organized into their own directories, 
 * we want to generate a test for each directory instead of
 * hardcoding them for project names--and having to update code
 * any time there is a change.
 * 
 * Note: This is very similar to the idea behind what Next.js's
 * getStaticPaths() method does for generating static pages.
 */
it('renders each ProjectsSection page', () => {
    // Retrieve all project directories
    const projects = getAllProjects();

    // Iterate over each project directory
    projects.map(proj => {
        // Retrieve the posts/content for each project
        const projectPosts = getPostsFromProjectsSubdir(proj);

        // Render the component, providing props for the iteration
        const { container } = render(
            <ProjectsSection 
                projectTitle={proj} 
                posts={projectPosts} />
        );
        expect(container).toMatchSnapshot();
    })
})