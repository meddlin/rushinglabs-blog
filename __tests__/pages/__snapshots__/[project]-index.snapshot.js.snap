// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders each ProjectsSection page 1`] = `
<div>
  <div>
    <header
      class="header"
    >
      <div
        class="menuIcon"
      >
        <div />
        <div />
        <div />
        
      </div>
      <h1
        class="heading2Xl"
      >
        <a
          class="colorInherit"
          href="/"
        >
          Rushing Labs
        </a>
      </h1>
      <div
        class="navbarLinks"
      >
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/blog"
          >
            Blog
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/projects"
          >
            Projects
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/docs"
          >
            Docs
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/about"
          >
            About
          </a>
        </h2>
      </div>
    </header>
    <div
      class="mainContent"
    >
      <div
        class="container"
      >
        <h1>
          Project: 
          amortize
        </h1>
        <ul />
      </div>
      <div
        class="backToHome"
      >
        <a
          href="/"
        >
          ← Back to home
        </a>
      </div>
      <a
        href="https://plausible.io/rushinglabs.com"
      >
        RushingLabs - Analytics
      </a>
    </div>
    
  </div>
</div>
`;

exports[`renders each ProjectsSection page 2`] = `
<div>
  <div>
    <header
      class="header"
    >
      <div
        class="menuIcon"
      >
        <div />
        <div />
        <div />
        
      </div>
      <h1
        class="heading2Xl"
      >
        <a
          class="colorInherit"
          href="/"
        >
          Rushing Labs
        </a>
      </h1>
      <div
        class="navbarLinks"
      >
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/blog"
          >
            Blog
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/projects"
          >
            Projects
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/docs"
          >
            Docs
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/about"
          >
            About
          </a>
        </h2>
      </div>
    </header>
    <div
      class="mainContent"
    >
      <div
        class="container"
      >
        <h1>
          Project: 
          game-capture
        </h1>
        <ul>
          <li>
            <a
              href="/projects/game-capture/intro"
            >
              Game Capture - a TikTok series
            </a>
          </li>
        </ul>
      </div>
      <div
        class="backToHome"
      >
        <a
          href="/"
        >
          ← Back to home
        </a>
      </div>
      <a
        href="https://plausible.io/rushinglabs.com"
      >
        RushingLabs - Analytics
      </a>
    </div>
    
  </div>
</div>
`;

exports[`renders each ProjectsSection page 3`] = `
<div>
  <div>
    <header
      class="header"
    >
      <div
        class="menuIcon"
      >
        <div />
        <div />
        <div />
        
      </div>
      <h1
        class="heading2Xl"
      >
        <a
          class="colorInherit"
          href="/"
        >
          Rushing Labs
        </a>
      </h1>
      <div
        class="navbarLinks"
      >
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/blog"
          >
            Blog
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/projects"
          >
            Projects
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/docs"
          >
            Docs
          </a>
        </h2>
        <h2>
          <a
            class="text-xl font-bold underline text-inherit"
            href="/about"
          >
            About
          </a>
        </h2>
      </div>
    </header>
    <div
      class="mainContent"
    >
      <div
        class="container"
      >
        <h1>
          Project: 
          owasp
        </h1>
        <ul>
          <li>
            <a
              href="/projects/owasp/a01-broken-access-control"
            >
              A01:2021 Broken Access Control
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a02-cryptographic-failures"
            >
              A02:2021 Cryptographic Failures
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a03-injection"
            >
              A03:2021 Injection
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a04-insecure-design"
            >
              A04:2021 Insecure Design
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a05-security-misconfiguration"
            >
              A05:2021 Security Misconfiguration
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a06-vulnerable-outdated-components"
            >
              A06:2021 Vulnerable and Outdated Components
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a07-identification-auth-failures"
            >
              A02:2021 Identification and Authentication Failures
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a08-software-data-integrity-failures"
            >
              A08:2021 Software and Data Integrity Failures
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a09-logging-monitoring-failures"
            >
              A09:2021 Security Logging and Monitoring Failures
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/a10-server-side-request-forgery"
            >
              A10:2021 Server Side Request Forgery (SSRF)
            </a>
          </li>
          <li>
            <a
              href="/projects/owasp/intro"
            >
              Intro: Why Bother "Explaning" the Top 10?
            </a>
          </li>
        </ul>
      </div>
      <div
        class="backToHome"
      >
        <a
          href="/"
        >
          ← Back to home
        </a>
      </div>
      <a
        href="https://plausible.io/rushinglabs.com"
      >
        RushingLabs - Analytics
      </a>
    </div>
    
  </div>
</div>
`;
